sns/parity_math.c

```diff
@@ -500,29 +551,114 @@ static void reed_solomon_encode(struct m0_parity_math *math,
 
 #undef PARITY_MATH_REGION_ENABLE
 }
+#endif /* __KERNEL__ */
+
+#ifndef __KERNEL__
+static void isal_encode(struct m0_parity_math *math,
+			const struct m0_buf *data,
+			struct m0_buf *parity)
+{
+	uint8_t **data_frags = NULL;
+	uint8_t **parity_frags = NULL;
+	uint32_t  i;
+	uint32_t  data_count;
+	uint32_t  parity_count;
+	uint32_t  block_size;
+	int	  ret = 0;
+
+	M0_PRE(math != NULL);
+	M0_PRE(data != NULL);
+	M0_PRE(parity != NULL);
+
+	M0_ENTRY();
+
+	data_count = math->pmi_data_count;
+	parity_count = math->pmi_parity_count;
+	block_size = data[0].b_nob;
+
+	M0_ALLOC_ARR(data_frags, data_count);
+	if (data_frags == NULL) {
+		ret = M0_ERR_INFO(-ENOMEM, "failed to allocate memory for "
+				  "array of data fragments");
+		goto fini;
+	}
+
+	M0_ALLOC_ARR(parity_frags, parity_count);
+	if (parity_frags == NULL) {
+		ret = M0_ERR_INFO(-ENOMEM, "failed to allocate memory for "
+				  "array of parity fragments");
+		goto fini;
+	}
+
+	data_frags[0] = (uint8_t *)data[0].b_addr;
+	for (i = 1; i < data_count; ++i) {
+		if (block_size != data[i].b_nob) {
+			ret = M0_ERR_INFO(-EINVAL, "data block size mismatch. "
+					  "block_size = %u, data[%u].b_nob=%u",
+					  block_size, i, data[i].b_nob);
+			goto fini;
+		}
+		data_frags[i] = (uint8_t *)data[i].b_addr;
+	}
+
+	for (i = 0; i < parity_count; ++i) {
+		if (block_size != parity[i].b_nob) {
+			ret = M0_ERR_INFO(-EINVAL, "parity block size mismatch. "
+					  "block_size = %u, parity[%u].b_nob=%u",
+					  block_size, i, parity[i].b_nob);
+			goto fini;
+		}
+		parity_frags[i] = (uint8_t *)parity[i].b_addr;
+	}
+
+	M0_LOG(M0_DEBUG, "initialize tables for fast Erasure Code encode.");
+	ec_init_tables(data_count, parity_count,
+		       &math->encode_matrix[data_count * data_count],
+		       math->g_tbls);
+
+	M0_LOG(M0_DEBUG, "generate erasure codes on given blocks of data.");
+	ec_encode_data(block_size, data_count, parity_count,
+		       math->g_tbls, data_frags, parity_frags);
+
+fini:
+	m0_free(data_frags);
+	m0_free(parity_frags);
+
+	/* TODO: Return error code instead of assert */
+	M0_ASSERT(ret == 0);
```

Comment:
Currently, this function returns void, so we ASSERT the ret code here.
In future, error code should be returned and errors should be handled by caller.

Reply:
Yes. We have created separate ticket [EOS-17288](https://jts.seagate.com/browse/EOS-17288) for the same.